applybn/feature_selection/bn_nmi_feature_selector.py

```diff
@@ -0,0 +1,130 @@
+
+from pyitlib import discrete_random_variable as drv
+import pandas as pd
+
+
+class NMIFeatureSelectior:
+  def __init__(self, threshhold = 0.5, bin_count=100, verboose=False):
+    self.threshhold = threshhold
+    self.bin_count = bin_count
+    self.verboose = verboose
+
+
+  def _discreticise(self, x, y=None):
+    '''
+    Prep data to feed into pyitlib drv.entropy
+    Calc happenings of each event in descreticised feature
+    '''
+    if y is None:
+      discr = [0] * (max(x) - min(x) + 1)
+      for e in x:
+        discr[int(e)] += 1
+      return discr
+
+    discr = dict()
+    for e1, e2 in zip(x, y):
+      if (e1,e2) not in discr:
+        discr[(e1,e2)] = 0
+      discr[(e1,e2)] += 1
+    lenx = len(self._discreticise(x)) + 1
+    leny = len(self._discreticise(y)) + 1
+    res = [0] * (lenx * leny)
+    for k, v in discr.items():
+      res[(k[0] * leny + k[1])] = v
+    return res
+
+
+  def _NMI(self, a, b):
+    '''
+    Calculate Normalized Mutual Information (NMI) for pair of features
+    '''
+    if (a.name, None) not in self.entropy_map:
+      self.entropy_map[(a.name, None)] = float(drv.entropy(self._discreticise(a)))
+    if (b.name, None) not in self.entropy_map:
+      self.entropy_map[(b.name, None)] = float(drv.entropy(self._discreticise(b)))
+    if (a.name, b.name) not in self.entropy_map:
+      self.entropy_map[(a.name, b.name)] = float(drv.entropy_joint(self._discreticise(a,b)))
+
+    Ha = self.entropy_map[(a.name, None)]
+    Hb = self.entropy_map[(b.name, None)]
+    Hab = self.entropy_map[(a.name, b.name)]
+    return (Ha + Hb - Hab) / (2 * max(Ha, Hb))
+
+  def _prep_dataset(self, df, target):
+    '''
+    Prepare the dataset, discretize the fields, remove rows
+    Return a ready-to-use dataset
+    '''
+    clean_df = pd.DataFrame()
+    for col in df:
+      if df[col].dtype in ['int', 'int32','int64']:
+        # integers can be discreticised this way, without binning
+        df_map = { l:i for i,l in enumerate(set(df[col].to_list()))}
+        clean_df[col] = df[col].map(df_map)
+      elif df[col].dtype in ['float', 'float64', 'double']:
+        # discretisize to bin_count bins
+        clean_df[col] = pd.cut(df[col].to_list(), bins=self.bin_count, labels=[i for i in range(self.bin_count)])
+        clean_df[col] = clean_df[col].cat.add_categories(self.bin_count + 1).fillna(self.bin_count + 1)
+      elif df[col].dtype in ['string', 'object']:
+        # pandas cut does not work with string features, this is done manually
+        # it is assumed that the user has removed the ID-like features, otherwise there will be a memory error later
+        df_map = { l:i for i,l in enumerate(set(df[col].to_list()))}
+        clean_df[col] = df[col].map(df_map)
+    return clean_df
+
+  def fit(self, df, target):
+    clean_df = self._prep_dataset(df, target)
```

Comment:
Depending on dataset size, repeated use of set(...) or enumerating large columns may be slow.
If large amounts of data are being processed, you might consider vectorization with NumPy or dedicated libs (like scikit-learn’s KBinsDiscretizer) for binning float columns.